cpp/src/arrow/csv/converter.cc

```diff
@@ -601,18 +631,30 @@ Result<std::shared_ptr<DictionaryConverter>> DictionaryConverter::Make(
   DictionaryConverter* ptr;
```

Comment:
Not really part of this patch, but: instead of having a raw pointer which is later upgraded to a shared pointer, please declare this as a smart pointer so it's always stack bound